### ISSUE TITLE
Library linking

### ISSUE THREAD

**Issue by klkvr** [foundry-rs/foundry#6215]
### Component

Forge

### Describe the feature you would like

Right now unused libraries may be deployed even if there is no usage of them outside of broadcasting:

- https://github.com/foundry-rs/foundry/issues/3295
- https://github.com/foundry-rs/foundry/issues/5375

This happens because from perspective of Solidity compiler, script contract depends on all libraries which are used in it regardeless of part of code they are used in.

One possible solution is as follows:
1. After script simulation, go through all `CREATE`'s and match bytecodes to artifacts.
2. Determine which libraries are actually going to be used in newly created contracts. This can be done via [existing code for dependency resolution
](https://github.com/foundry-rs/foundry/blob/master/crates/utils/src/lib.rs)
3. If any unnecessary libraries are found, recompile the code and rerun the script, predeploying only the necessary libraries. This step is necessary because modifying the pre-deployed libraries list affects the sender nonce. All other libraries can be injected by some way similar to what `vm.etch` does

This approach has several drawbacks:
1. Current code matching bytecode to artifacts uses [diff_score](https://github.com/foundry-rs/foundry/blob/master/crates/common/src/contracts.rs#L25) which may result in false positives/negatives in some cases. Right now such fails are accepted because they are only affecting verification and trace decoding. However, in case of library linking it becomes much more dangerous as it may lead to contracts being deployed without needed libraries
2. Recompiling code and re-running script with new libraries is pretty expensive.

Would be happy to get other opinions on this

**Comment by mattsse** [comment]
yeah this entire thing is giga cursed atm...

this needs a proper rewrite

**Comment by klkvr** [comment]
yep it's still an issue. basically if script depends on any libraries all of them are going to be deployed when it's executed, even if they are used outside of broadcast blocks.

this can be worked around by isolating scripts which depend on such contracts

I didn't figure out a nice solution for this yet which wouldn't require re-executing the script :/

**Comment by klkvr** [comment]
something I've considered which seems a bit hacky is the following:
1. Before execution link script with dummy library addresses (just random 20-byte arrays which are easy to identify). Keep a reference from dummy address to library it represents.
2. During execution, if we see a `CREATE` with bytecode referencing dummy address generated above, we firstly push a transaction for deployment of this library to `broadcastable_transactions`. That way, we will only deploy libraries which broadcasted creation transactions depend on

**Comment by xrave110** [comment]
Is it the way to use libraries deployed in previous script ? I have list of scripts that needs to be performed in order and now each script from the list deploys the same libraries every time.
